### PR TITLE
Return directory path without a trailing DS.

### LINKED ISSUE #45: resolve must return a path with a trailing DS only when it's a directory
Please, see the discussion here https://github.com/hoaproject/Http/pull/4.

### PR DESCRIPTION
**Pull request by Metalaka**
Fix #45 .

### CODE REVIEW

**Comment by Metalaka** [comment]
/ping
We must normalize this behavior, in a case like in the other (with or without a trailing DS).
Personally, I prefer without the DS (see @jubianchi's arguments).

Should we do something like a vote ?

**Comment by Jir4** [comment]
I agree with @jubianchi (and @Metalaka so)

**Comment by Savageman** [comment]
As a PHP developper, I expect that directory don't have a traling /, expect for the root directory. That's how it works in this language and what is described in the official documentation.

**Comment by shouze** [comment]
@ping everyone. In computer science there's already conventions about paths. I think this [wikipedia article](http://en.wikipedia.org/wiki/Path_(computing)) cover pretty well the subject.

We learn (or remind) it exists 2 standards:
- UNC
- POSIX

Both of them agree that:
- Trailing DS are not necessary even if not explicitly prohibited.
- Redundant & contiguous DS are treated like unique ones.

I'm not in favor to add some trailing DS at all.

**Comment by Hywan** [comment]
I am also in favor to add a trailing DS.

/ping @hoaproject/hoackers for the vote :-).

**Comment by shouze** [comment]
:-1:

**Comment by stephpy** [comment]
This is not to Hoa to change the way systems works.

Totally against trailing DS.

**Comment by vonglasow** [comment]
```php
$path = …;
$file = $path . '/foo';
unlink($file);
```
@Hywan Even if you have `$path` empty if you have a trailing / you will have a `$file = //foo` and generaly it works as `/foo` or we use trim to remove trailing /
In other way, I hate having to add a / in requirement like
```php
require_once $path . '/foo';
```
But I am agree with @jubianchi and even if I prefer to have a trailing / I think is more "standard" to not add it.

**Comment by shulard** [comment]
I've encountered an issue about that trailing `/` dilemma. I think that resolve mustn't return a trailing `/` if the path is only a directory or just a resolvable one (like `hoa://Data/Variable/Run`).

I've tried playing with the Worker lib and the path resolution give me some trouble...

Override `hoa://Data/Variable/Run` with a custom path (note that I do not add a trailing `/` to the new path): 
```php
Hoa\Core\Core::getInstance()->setProtocol(
  'hoa://Data/Variable/Run',
  "\r".'/foo/bar/wid'
);
```

The `wid` files get created at : `/foo/bar/widdemorker.wid`. I think it must have been created at `/foo/bar/wid/demorker.wid` because I gave a valid folder in the `setProtocol` method.

With the new path `"\r".'/foo/bar/wid/'` the file is created properly. I check the Worker lib and it use the resolve method to build the `wid` file path...

Maybe `setProtocol` must handle the trailing `/`, maybe the resolve must handle it but I think that it must be managed somewhere here.

**Comment by Metalaka** [comment]
The previous vote is in favor of don't add trailing DS.
So, I updated my PR to follow this but all `resolve` uses must be check because the default behavior has changed.

**Comment by Hywan** [comment]
Excellent, good :-). I am going to focus on `Hoa\Core` for the next days.

**Comment by Metalaka** [comment]
Surprised that there is only few call to `resolve` in all libraries.
Moreover, it seem there is only one call which need to be fixed.

**Comment by Hywan** [comment]
@Metalaka Most of the time, we don't need `resolve`, just calling `hoa://` in a stream, resource etc. work because it's very low-level in the PHP land. We call `resolve` when we need the final path, not the `hoa://` path (it happens sometimes).

File: Protocol.php [hunk 1]
```diff
@@ -201,6 +201,12 @@ public function offsetUnset($name)
     public function resolve($path, $exists = true, $unfold = false)
     {
         if (substr($path, 0, 6) !== 'hoa://') {
+            if (is_dir($path)) {
+                $path = rtrim($path, '/\\');
```
Comment (Hywan) [inline Protocol.php]: Add a new line after `$path …;` please.

File: Protocol.php [hunk 2]
```diff
@@ -216,6 +222,15 @@ public function resolve($path, $exists = true, $unfold = false)
 
             $handle = array_values(array_unique($handle, SORT_REGULAR));
 
+            foreach ($handle as &$entry) {
+                if (is_dir($entry)) {
+                    $entry = rtrim($entry, '/\\');
```
Comment (Hywan) [inline Protocol.php]: Add a new line after `$entry =…` please.

File: Protocol.php [hunk 3]
```diff
@@ -201,6 +201,12 @@ public function offsetUnset($name)
     public function resolve($path, $exists = true, $unfold = false)
     {
         if (substr($path, 0, 6) !== 'hoa://') {
+            if (is_dir($path)) {
```
Comment (Hywan) [inline Protocol.php]:
```php
if (true === is_dir(…))
```

File: Protocol.php [hunk 4]
```diff
@@ -216,6 +222,15 @@ public function resolve($path, $exists = true, $unfold = false)
 
             $handle = array_values(array_unique($handle, SORT_REGULAR));
 
+            foreach ($handle as &$entry) {
+                if (is_dir($entry)) {
```
Comment (Hywan) [inline Protocol.php]:
```php
if (true === is_dir(…))
```

**Comment by Hywan** [comment]
@Metalaka Good to merge?

**Comment by Metalaka** [comment]
Yep, It's ready.

**Comment by Hywan** [comment]
https://github.com/hoaproject/Core/commit/0e7a331a66bdbbd0eaae4fee8da196ad26b0cd95 merged.
https://github.com/hoaproject/Core/commit/54f1c4eac817de324fadff0d1a1322d2328b5497 merged.

(I run a `git rebase master`).

Thanks!

**Pull request closed without merging**